09-typescript-in-react/README.md

```diff
@@ -0,0 +1,641 @@
+    return;
+  }
+  //...
+};
+```
+
+#### Typing Async Functions
+
+`async` and `await` are just a syntax sugar over the Promises, that's why we can type the async function like this:
+
+```TypeScript
+type Data = {
+  id: number;
+  info: string;
+};
+
+const fetchData = async(query: string): Promise<Data[]> => { //<= Promise that will resolve to array of Data objects
+  const result = await fetch(`https://api.domain.com/data?query=${query}`);
+  const resultJSON = await result.json();
+  return resultJSON.result || [];
+}
+```
+
+#### Default Parameters
+
+It is possible to use default parameters in the function declaration. Using default parameters will make them optional:
+
+```TypeScript
+function sayMyName(name: string = 'Fantomas'): void {
+  // ...
+};
+```
+
+There is no need to use question marks for parameters with default values.
+
+### Generics
+
+Thanks to generics, it is possible to create functions and classes that work with a variety of types. For example instead
+of having:
+
+```TypeScript
+function doSomethingWithString(str: string): string {
+  return str;
+}
+
+function doSomethingWithNumber(num: number): number {
+  return num;
+}
+```
+
+we use generic type and define the `doSomething` method like this:
+
+```TypeScript
+function doSomething<T>(val: T): T {
+  return val;
+}
+```
+
+Then we can call the function either with type in brackets as parameter:
+
+```TypeScript
+const result = doSomething<string>('Hello');
+```
+
+Or we can even rely on TypeScript that it will infer the type and simply call:
+
+```TypeScript
+const result = doSomething('Hello');
+```
+
+Variable `result` will be a string in both cases.
+
+We can define type alias for function with generics e.g. like this:
+
+```TypeScript
+type MyGenericFunction = <T>(arg: T) => T;
+```
+
+Generics can be utilized also in classes:
+
+```TypeScript
+class GenericClass<T> {
+  private myValue: T;
+
+  constructor (myValue: T) {
+    this.myValue = myValue;
+  }
+
+  public getValue(): T {
+    return this.myValue;
+  };
+}
+
+const MyClass = new GenericClass<number>(4);
+const result = MyClass.getValue();
+// type of result is number
+```
+
+### Useful Utility Types
+
+Utility types are generic types that we can use out of the box and that transforms given type somehow.
+
+#### Partial
+
+```TypeScript
+type Car = {
+  color: string;
+  maxSpeed: number;
+};
+
+type PartialCar = Partial<Car>;
+// type PartialCar = {
+//   color?: string;
+//   maxSpeed?: number;
+// };
+```
+
+#### Pick
+
+```TypeScript
+type Car = {
+  color: string;
+  maxSpeed: number;
+};
+
+type PickCar = Pick<Car, 'color'>;
+// type PickCar = {
+//   color: string;
+// };
+```
+
+#### Omit
+
+```TypeScript
+type Car = {
+  color: string;
+  maxSpeed: number;
+};
+
+type OmitCar = Omit<Car, 'color'>;
+// type OmitCar = {
+//   maxSpeed: number;
+// };
+```
+
+More utility types can be found [here](https://www.typescriptlang.org/docs/handbook/utility-types.html).
+
+### Usage in React
```

Comment:
not sure how much time you want to spend on the theory above,, IMHO you could ask people if they want to see `Usage in React` before going into too much details :)

Reply:
Yeah... I am also afraid that the theory might be long, but I am trying to explain fundamental concepts they should know. Not sure how to do it - if I jump right into react without explaining generics and fundamental concepts, they might be confused. All depends on the audience and their experience - if they know TS or other strongly typed language.